### PR TITLE
Fix scope for variables introduced by pattern expressions

### LINKED ISSUE #4525: Pattern variables not always introduced into the correct scope
https://docs.oracle.com/javase/specs/jls/se21/html/jls-6.html#jls-6.3.1 and onwards in the language specification provides a list of rules for how variables are introduced into various scopes. The JavaParser symbol solver handles many of these correctly, but doesn't implement some of these rules correctly. I've implemented tests for all of the rules in the spec [here](https://github.com/johannescoetzee/javaparser/blob/969555f3c3c41707916b38f079cd30867d11d8be/javaparser-symbol-solver-testing/src/test/java/com/github/javaparser/symbolsolver/resolution/PatternVariableIntroductionTest.java), but include the failing testcases below. I've double checked that these code examples compile, so these variables should all be resolvable.

I'm going to try to fix this, but decided to open this issue for documentation and to discuss ideas.

## Failing tests:
```
@Test
public void ifIntroducesPatternVariable1() {
    CompilationUnit cu = parse("public class Test {\n"
            + "  public void foo(Object o) {\n"
            + "    if (!(o instanceof String s)) {\n"
            + "      return;\n"
            + "    }\n"
            + "    System.out.println(s);\n"
            + "  }\n"
            + "}");

    NameExpr name = Navigator.findNameExpression(cu, "s").get();
    assertEquals("java.lang.String", name.resolve().getType().describe());
}
```

```
@Test
public void ifIntroducesPatternVariable7() {
    CompilationUnit cu = parse("public class Test {\n"
            + "  public void foo(Object o, boolean b) {\n"
            + "    if (!(o instanceof String s)) {\n"
            + "        return;\n"
            + "    } else if (b) {\n"
            + "        // do nothing\n"
            + "    } \n"
            + "    System.out.println(s);\n"
            + "  }\n"
            + "}");

    NameExpr name = Navigator.findNameExpression(cu, "s").get();
    assertEquals("java.lang.String", name.resolve().getType().describe());
}
```

```
@Test
public void ifIntroducesPatternVariable8() {
    CompilationUnit cu = parse("public class Test {\n"
            + "  public void foo(Object o, boolean b) {\n"
            + "    if (!(o instanceof String s)) {\n"
            + "        return;\n"
            + "    } else if (b) {\n"
            + "        // do nothing\n"
            + "    } else {}\n"
            + "    System.out.println(s);\n"
            + "  }\n"
            + "}");

    NameExpr name = Navigator.findNameExpression(cu, "s").get();
    assertEquals("java.lang.String", name.resolve().getType().describe());
}
```

```
@Test
public void ifIntroducesPatternVariable9() {
    CompilationUnit cu = parse("public class Test {\n"
            + "  public void foo(Object o, boolean b) {\n"
            + "    if (!(o instanceof String s)) {\n"
            + "        return;\n"
            + "    } else if (b) {\n"
            + "        // do nothing\n"
            + "    } else {\n"
            + "        return;\n"
            + "    }\n"
            + "    System.out.println(s);\n"
            + "  }\n"
            + "}");

    NameExpr name = Navigator.findNameExpression(cu, "s").get();
    assertEquals("java.lang.String", name.resolve().getType().describe());
}
```

```
@Test
public void conditionalOperator1() {
    CompilationUnit cu = parse("public class Test {\n"
            + "    public void foo(Object o) {\n"
            + "      boolean b = o instanceof String value ? value.length() > 0 : false;\n"
            + "    }\n"
            + "}");

    NameExpr name = Navigator.findNameExpression(cu, "value").get();
    assertEquals("java.lang.String", name.resolve().getType().describe());
}
```

```
@Test
public void conditionalOperator2() {
    CompilationUnit cu = parse("public class Test {\n"
            + "    public void foo(Object o) {\n"
            + "      boolean b = !(o instanceof String value) ? false : value.length() > 5;\n"
            + "    }\n"
            + "}");

    NameExpr name = Navigator.findNameExpression(cu, "value").get();
    assertEquals("java.lang.String", name.resolve().getType().describe());
}
```

```
@Test
public void conditionalOr1() {
    CompilationUnit cu = parse("public class Test {\n"
            + "    public void foo(Object o) {\n"
            + "        if (!(o instanceof String value) || value.length() > 5) { }\n"
            + "    }\n"
            + "}");

    NameExpr name = Navigator.findNameExpression(cu, "value").get();
    assertEquals("java.lang.String", name.resolve().getType().describe());
}
```

```
@Test
public void conditionalOr2_1() {
    CompilationUnit cu = parse("public class Test {\n"
            + "    public void foo(Object o, boolean b) {\n"
            + "        if (!(!(o instanceof String value) || b)) {\n"
            + "            System.out.println(value);\n"
            + "        }\n"
            + "    }\n"
            + "}");

    NameExpr name = Navigator.findNameExpression(cu, "value").get();
    assertEquals("java.lang.String", name.resolve().getType().describe());
}
```

```
@Test
public void conditionalOr2_2() {
    CompilationUnit cu = parse("public class Test {\n"
            + "    public void foo(Object o, boolean b) {\n"
            + "        if (!(b || !(o instanceof String value))) {\n"
            + "            System.out.println(value);\n"
            + "        }\n"
            + "    }\n"
            + "}");

    NameExpr name = Navigator.findNameExpression(cu, "value").get();
    assertEquals("java.lang.String", name.resolve().getType().describe());
}
```

```
@Test
public void doStatements1() {
    CompilationUnit cu = parse("class Test {\n"
            + "    public void test(Object o) {\n"
            + "        do {\n"
            + "        } while (!(o instanceof String value));\n"
            + "        System.out.println(value);\n"
            + "    }\n"
            + "}");

    NameExpr name = Navigator.findNameExpression(cu, "value").get();
    assertEquals("java.lang.String", name.resolve().getType().describe());
}
```

```
@Test
public void forStatements1() {
    CompilationUnit cu = parse("class Test {\n"
            + "    public void test(Object o) {\n"
            + "        for (int i = 0; o instanceof String value; i += value.length()) { }\n"
            + "    }\n"
            + "}");

    NameExpr name = Navigator.findNameExpression(cu, "value").get();
    assertEquals("java.lang.String", name.resolve().getType().describe());
}
```

```
@Test
public void forStatements2() {
    CompilationUnit cu = parse("class Test {\n"
            + "    public void test(Object o) {\n"
            + "        for (; o instanceof String value; ) {\n"
            + "            System.out.println(value);\n"
            + "        }\n"
            + "    }\n"
            + "}");

    NameExpr name = Navigator.findNameExpression(cu, "value").get();
    assertEquals("java.lang.String", name.resolve().getType().describe());
}
```

```
@Test
public void forStatements3() {
    CompilationUnit cu = parse("class Test {\n"
            + "    public void test(Object o) {\n"
            + "        for (; !(o instanceof String value); ) {\n"
            + "        }\n"
            + "        System.out.println(value);\n"
            + "    }\n"
            + "}");

    NameExpr name = Navigator.findNameExpression(cu, "value").get();
    assertEquals("java.lang.String", name.resolve().getType().describe());
}
```

```
@Test
public void ifStatements2() {
    CompilationUnit cu = parse("class Test {\n"
            + "    public void test(Object o) {\n"
            + "        if (!(o instanceof String value)) {\n"
            + "            return;\n"
            + "        }\n"
            + "        System.out.println(value);\n"
            + "    }\n"
            + "}");

    NameExpr name = Navigator.findNameExpression(cu, "value").get();
    assertEquals("java.lang.String", name.resolve().getType().describe());
}
```

```
@Test
public void ifStatements3() {
    CompilationUnit cu = parse("class Test {\n"
            + "    public void test(Object o) {\n"
            + "        if (!(o instanceof String value)) {\n"
            + "            throw new RuntimeException();\n"
            + "        }\n"
            + "        System.out.println(value);\n"
            + "    }\n"
            + "}");

    NameExpr name = Navigator.findNameExpression(cu, "value").get();
    assertEquals("java.lang.String", name.resolve().getType().describe());
}
```

```
@Test
public void ifStatements6() {
    CompilationUnit cu = parse("class Test {\n"
            + "    public void test(Object o) {\n"
            + "        if (o instanceof String value) {\n"
            + "        } else {\n"
            + "            return;\n"
            + "        }\n"
            + "        System.out.println(value);\n"
            + "    }\n"
            + "}");

    NameExpr name = Navigator.findNameExpression(cu, "value").get();
    assertEquals("java.lang.String", name.resolve().getType().describe());
}
```

```
@Test
public void ifStatements7() {
    CompilationUnit cu = parse("class Test {\n"
            + "    public void test(Object o) {\n"
            + "        if (!(o instanceof String value)) {\n"
            + "            return;\n"
            + "        } else {\n"
            + "        }\n"
            + "        System.out.println(value);\n"
            + "    }\n"
            + "}");
    NameExpr name = Navigator.findNameExpression(cu, "value").get();
    assertEquals("java.lang.String", name.resolve().getType().describe());
}
```

```
@Test
public void whileStatements1() {
    CompilationUnit cu = parse("class Test {\n"
            + "    public void test(Object o) {\n"
            + "        while (o instanceof String value) {\n"
            + "            System.out.println(value);\n"
            + "        }\n"
            + "    }\n"
            + "}");

    NameExpr name = Navigator.findNameExpression(cu, "value").get();
    assertEquals("java.lang.String", name.resolve().getType().describe());
}
```

```
@Test
public void whileStatements2() {
    CompilationUnit cu = parse("class Test {\n"
            + "    public void test(Object o) {\n"
            + "        while (!(o instanceof String value)) {\n"
            + "        }\n"
            + "        System.out.println(value);\n"
            + "    }\n"
            + "}");

    NameExpr name = Navigator.findNameExpression(cu, "value").get();
    assertEquals("java.lang.String", name.resolve().getType().describe());
}
```

```
@Test
public void whileStatements3() {
    CompilationUnit cu = parse("class Test {\n"
            + "    public void test(Object o) {\n"
            + "        first:\n"
            + "        while (true) {\n"
            + "            while (!(o instanceof String value)) {\n"
            + "                break first;\n"
            + "            }\n"
            + "            System.out.println(value);\n"
            + "        }\n"
            + "    }\n"
            + "}");

    NameExpr name = Navigator.findNameExpression(cu, "value").get();
    assertEquals("java.lang.String", name.resolve().getType().describe());
}
```

### PR DESCRIPTION
**Pull request by johannescoetzee**
Fixes #4525 

I tried to stay as close as possible to the JLS for the changes in this PR since the rules regarding variables introduced by pattern expressions are very specific. For most of the methods I added, I added a comment quoting the rule from the JLS that is implemented by the specific method. If something doesn't make sense, feel free to ask and I'll try to explain (or correct it if I made a mistake) :)

# PatternVariableVisitor
The `PatternVariableVisitor` was added to replace the `*typePatternExprsExposedFromChildren` methods in `Context`. In my opinion this is a cleaner implementation for 2 reasons:
* It avoids having to create contexts for all the children as the tree is traversed downwards looking for patterns
* It keeps the logic for searching for patterns contained in a single file which makes it easier to read through and reason about.

# Context changes
A new `getIntroducedTypePatterns` method was added to `StatementContext` with implementations for `if`, `do`, `while`, and `for` statements. This method is used to find pattern expressions introduced by rules such as
```
A pattern variable is introduced by if (e) S iff (i) it is introduced by e when false and (ii) S cannot complete normally. 
```

When attempting to solve a type in context, one of the steps is now to search for pattern expressions exposed to the child in the parent. I implemented it this way instead of attempting to solve the symbol in the parent context directly since this avoids the problem mentioned in https://github.com/javaparser/javaparser/issues/4344, among others. This adds yet another path through which a symbol can be solved, which is not ideal, but fixing the issues with losing positional information/context history when attempting to solve symbols is outside the scope of this PR (although I'd be happy to take a look at some point since I think I understand the `javaparsermodel` process for solving symbols fairly well by now).

I've also cleaned up implementations for the `*typePatternExprsExposedFromChildren` methods and removed now-unnecessary helper methods.

# NormalCompletionVisitor
Some rules (such as that for if statements mentioned above) require knowledge about whether a statement can complete normally. Similar to the PatternVariableVisitor, I've added a separate visitor for this to keep the logic contained. I've implemented most of the rules described in the JLS, but notably did not implement the rules for loops correctly since this requires knowledge of whether the loop condition is a constant expression that evaluates to true/false. Since this PR was already getting quite big and this only affects an edge-case which I suspect is rather rare, I decided to put it aside for now with a comment of what needs to change. I'll come back to it at some point, but I think that's worth a separate change.

# Test changes
I added tests to test all of the rules for pattern variable scope mentioned in the JLS and also removed some old tests that were testing incorrect behaviour (see the linked issue for a comment describing some of these rules). I verified all of the tests that I've added with javac.

I added some negative tests of the form
```
        @Test
        public void conditionalOr1Negated() {
            CompilationUnit cu = parse("public class Test {\n"
                    + "    public void foo(Object o) {\n"
                    + "        if (o instanceof String value || value.length() > 5) { }\n"
                    + "    }\n"
                    + "}");

            NameExpr name = Navigator.findNameExpression(cu, "value").get();
            assertThrows(UnsolvedSymbolException.class, () -> name.resolve());
        }
```
Even though the code in this test doesn't compile, these tests still confirm that a pattern variable isn't incorrectly introduced into a given scope. As mentioned in the linked issue, one can always add a field `value` which would make this example compile and which would demonstrate how this could cause shadowing issues, but the tests as I wrote them are easier to verify with the compiler and test only how pattern variables are introduced into scope, instead of a combination of how the variables are introduced into scope and resolution of fields, for example.

### CODE REVIEW

File: javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/PatternVariableVisitor.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+import com.github.javaparser.ast.visitor.GenericVisitorWithDefaults;
+import java.util.*;
+
+public class PatternVariableVisitor extends GenericVisitorWithDefaults<PatternVariableResult, Void> {
+
+    private static PatternVariableVisitor patternVariableVisitor = null;
+
+    private PatternVariableVisitor() {}
+
+    public static PatternVariableVisitor getInstance() {
```
Comment (jlerbsc) [inline javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/PatternVariableVisitor.java]: Why create a singleton?

**Comment by jlerbsc** [comment]
It's a great job. Let me know when this PR is ready to be integrated. I don't have the knowledge to validate it fully, but if the tests pass and the existing system continues to work, I'm not opposed to integrating it.

File: javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/PatternVariableVisitor.java (hunk 1 above)
Comment (johannescoetzee) [reply javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/PatternVariableVisitor.java]: My reasoning was that there's no reason not to do it since it doesn't hold any state that would cause issues and by having it be a singleton, we skip the cost of creating a new instance with every symbol resolution. It's largely a stylistic choice though since I don't think it would have a noticeable performance impact to create new instances when needed. I'd be happy to refactor this if it's inconsistent with the rest of the project

**Comment by johannescoetzee** [comment]
If the singleton comment above isn't a blocker, then I think this should be ready to merge. The various paths symbol resolution can take get quite complex, so I can't say for sure that this won't include any regressions, but I did make sure to check that the existing tests that failed because of the behaviour change (which I then removed) were actually testing the incorrect behaviour. If any issues do come up that look like they may be related to this, please @ me in the issue and I'll take a look :)

File: javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/PatternVariableVisitor.java (hunk 1 above)
Comment (jlerbsc) [reply javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/PatternVariableVisitor.java]: There is no state in this visitor, but the cost of creating the visitor occasionally has no impact on performance. Nor is this a widespread pattern in JP. You can replace this pattern by creating an instance of the visitor.

Comment (johannescoetzee) [reply javaparser-symbol-solver-core/src/main/java/com/github/javaparser/symbolsolver/javaparsermodel/PatternVariableVisitor.java]: Done!
